vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/TextField.java

```diff
@@ -81,6 +83,8 @@ private TextField(boolean isInitialValueOptional) {
         setValueChangeMode(ValueChangeMode.ON_CHANGE);
 
         addValueChangeListener(e -> validate());
+
+        addClientValidatedEventListener(e -> validate());
```

Comment:
There's still one missing piece, which is to fire the `ValidationStatusChangeEvent` when the client side validation changes. Testing how it works, it seems that once the event is dispatched, the binder will take the validation status sent on the event to set the component is valid/invalid. So, the `getDefaultValidator()` is not called.

The thing I am still thinking about is how the `validate` method comes into play here. It needs to be called so that the component being used without binder will get the invalid state correctly set. But when binder is being used, we would probably need to call `getValidationSupport().isInvalid(getValue())` to get the validation result and send it to binder with the event.

Reply:
Good point. My first impression of `ValidationStatusChangeEventListener` was that we will need to subscribe **the binder** to `ClientValidatedEvent` by adding `ValidationStatusChangeEventListener` on `ClientValidatedEvent`.

Reply:
So, the final flow will be the following:

1. `ValueChange` triggers constraint validation and then binder validation.
2. `ClientValidated` triggers constraint validation and then binder validation by invoking `ValidationStatusChangeEventListener` listeners.

Reply:
I actually had a wrong assumption here:

> So, the getDefaultValidator() is not called.

My mistake was because I placed the breakpoint on the method body instead of the lambda's. What I can understand now is:
1. Client side validation fires `validated` event
2. Flow component listens the event and fires `ValidationStatusChangeEventListener`
3. Binder listens to this event and calls its `validate()` method
4. Eventually, `validate()` method will call the lambda created on `getDefaultValidator()`.
5. Component's `checkValidity()` method is called.